src/nbl/asset/utils/ISPIRVDebloater.cpp

```diff
@@ -45,71 +114,131 @@ ISPIRVDebloater::Result ISPIRVDebloater::debloat(const  ICPUBuffer* spirvBuffer,
             case  spv::ExecutionModelCallableKHR : return hlsl::ESS_CALLABLE;
             default:
             {
-                assert(false);
                 return hlsl::ESS_UNKNOWN;
             }
         }
       };
+
     static constexpr auto HEADER_SIZE = 5;
 
-    core::vector<bool> foundEntryPoints(entryPoints.size(), false);
     std::vector<uint32_t> minimizedSpirv;
-    bool needDebloat = false;
+    core::unordered_set<uint32_t> removedEntryPointIds;
 
+    bool needDebloat = false;
     auto offset = HEADER_SIZE;
+    auto parse_instruction = [](uint32_t instruction) -> std::tuple<uint32_t, uint32_t>
+    {
+        const auto length = instruction >> 16;
+        const auto opcode = instruction & 0x0ffffu;
+        return { length, opcode };
+    };
 
+    // skip until entry point
+    while (offset < spirvDwordCount) {
+        const auto instruction = spirv[offset];
+        const auto [length, opcode] = parse_instruction(instruction);
+        if (opcode == spv::OpEntryPoint) break;
+        offset += length;
+    }
+
+    // handle entry points removal
     while (offset < spirvDwordCount) {
         const auto curOffset = offset;
         const auto instruction = spirv[curOffset];
+        const auto [length, opcode] = parse_instruction(instruction);
+        if (opcode != spv::OpEntryPoint) break;
```

Comment:
what's the logic behind breaking out on the very first instruction that's not an entry point? shouldn't this be a `continue` ?

Reply:
also done after advancing `offset += length`